### PR TITLE
tests: drivers: adc-dma: fix the dma dest address check failure

### LINKED ISSUE #49792: test-ci:  adc-dma :  frdm-k64f: dma dest addess assert
**Describe the bug**
dma driver meet address assert due to address is not 32 bit aligned.

Please also mention any information which could help others to understand
the problem you're facing:

```
 ASSERTION FAIL [(((uint32_t)(uint32_t *)destAddr) % destWidth) == 0U] @ WEST_TOPDIR/modules/hal/nxp/mcux/mcux-sdk/drivers/edma/fsl_edma.c:1015
```

**To Reproduce**
cd tests/driver/adc/adc_dma
1. mkdir build; cd build
2. west build -b frdm_k64f
3. west flash

**Expected behavior**
test pass

**Impact**
adc dma

**Logs and console output**

```
*** Booting Zephyr OS build zephyr-v3.1.0-4370-g4f8156828674  ***
    Running TESTSUITE adc_dma
    ===================================================================
    START - test_adc_asynchronous_call
    ASSERTION FAIL [(((uint32_t)(uint32_t *)destAddr) % destWidth) == 0U] @ WEST_TOPDIR/modules/hal/nxp/mcux/mcux-sdk/drivers/edma/fsl_edma.c:1015
    E: r0/a1:  0x00000004  r1/a2:  0x000003f7  r2/a3:  0x000000a2
    E: r3/a4:  0x0000000f r12/ip:  0x00000000 r14/lr:  0x00006cf3
    E:  xpsr:  0x4100000f
    E: Faulting instruction address (r15/pc): 0x0000e52c
    E: >>> ZEPHYR FATAL ERROR 4: Kernel panic on CPU 0
    E: Fault during interrupt handling
    E: Current thread: 0x200004a0 (test_adc_asynchronous_call)
    E: Halting system
```

**Environment (please complete the following information):**
 - OS: (e.g. Linux,)
 - Toolchain (e.g Zephyr SDK, ...)
 - Commit SHA or Version used:  zephyr-v3.1.0-4370-g4f8156828674

### PR DESCRIPTION
**Pull request by hakehuang**
force the dma destination address 32 bit aligned

Fixes #49792

Signed-off-by: Hake Huang <hake.huang@oss.nxp.com>

### CODE REVIEW

File: tests/drivers/adc/adc_dma/src/test_adc.c [hunk 1]
```diff
@@ -37,8 +37,8 @@
 #define SAMPLE_INTERVAL_US (10000U)
 
 #define BUFFER_SIZE 24
-static ZTEST_BMEM int16_t m_sample_buffer[BUFFER_SIZE];
-static ZTEST_BMEM int16_t m_sample_buffer2[2][BUFFER_SIZE];
+static ZTEST_BMEM __aligned(32)  int16_t m_sample_buffer[BUFFER_SIZE];
+static ZTEST_BMEM __aligned(32)  int16_t m_sample_buffer2[2][BUFFER_SIZE];
```
Comment (stephanosio) [inline tests/drivers/adc/adc_dma/src/test_adc.c:41]: why two spaces?

```suggestion
static ZTEST_BMEM __aligned(32) int16_t m_sample_buffer[BUFFER_SIZE];
static ZTEST_BMEM __aligned(32) int16_t m_sample_buffer2[2][BUFFER_SIZE];
```

**Review by stephanosio — DISMISSED** [review]
commit title should start with `tests: drivers: adc_dma: ...`.

**Comment by teburd** [comment]
there is a dma alignment attribute in device tree now that could be used for this rather than hard coding to 32

File: tests/drivers/adc/adc_dma/src/test_adc.c [hunk 2]
```diff
@@ -37,8 +37,8 @@
 #define SAMPLE_INTERVAL_US (10000U)
 
 #define BUFFER_SIZE 24
-static ZTEST_BMEM int16_t m_sample_buffer[BUFFER_SIZE];
-static ZTEST_BMEM int16_t m_sample_buffer2[2][BUFFER_SIZE];
+static ZTEST_BMEM __aligned(32)  int16_t m_sample_buffer[BUFFER_SIZE];
```
Comment (teburd) [inline tests/drivers/adc/adc_dma/src/test_adc.c:40]: Please use the newly added dma alignment device tree property for this rather than hard coding it.

**Review by teburd — CHANGES_REQUESTED** [review]

File: tests/drivers/adc/adc_dma/src/test_adc.c (hunk 1 above)
Comment (hakehuang) [reply tests/drivers/adc/adc_dma/src/test_adc.c:41]: Thanks I will update.

File: tests/drivers/adc/adc_dma/src/test_adc.c (hunk 2 above)
Comment (hakehuang) [reply tests/drivers/adc/adc_dma/src/test_adc.c]: @teburd, Thanks I have update with dts alignment

File: tests/drivers/adc/adc_dma/boards/frdm_k64f.overlay [hunk 3]
```diff
@@ -10,3 +10,9 @@
 	high-speed;
 	periodic-trigger;
 };
+
+&edma0 {
+	dma-buf-alignment = <32>;
```
Comment (teburd) [inline tests/drivers/adc/adc_dma/boards/frdm_k64f.overlay:15]: This should be in the soc dtsi not a test overlay so other people can use it

File: tests/drivers/adc/adc_dma/boards/frdm_k82f.overlay [hunk 4]
```diff
@@ -10,3 +10,9 @@
 	high-speed;
 	periodic-trigger;
 };
+
+&edma0 {
+        dma-buf-alignment = <32>;
```
Comment (teburd) [inline tests/drivers/adc/adc_dma/boards/frdm_k82f.overlay:15]: same as above

**Review by teburd — DISMISSED** [review]
Closer but I think the alignment attribute is being put in the wrong spot here

File: tests/drivers/adc/adc_dma/boards/frdm_k64f.overlay (hunk 3 above)
Comment (hakehuang) [reply tests/drivers/adc/adc_dma/boards/frdm_k64f.overlay:15]: @teburd , for nxp edma, the buffer alignment is kind of variable 32 bit is the maximum requirement, see hal code below
```
assert((srcWidth != 0U) && (srcWidth <= 32U) && ((srcWidth & (srcWidth - 1U)) == 0U));
    assert((destWidth != 0U) && (destWidth <= 32U) && ((destWidth & (destWidth - 1U)) == 0U));
    assert((transferBytes % bytesEachRequest) == 0U);
    assert((((uint32_t)(uint32_t *)srcAddr) % srcWidth) == 0U);
    assert((((uint32_t)(uint32_t *)destAddr) % destWidth) == 0U);
``` 
it requires the destAddr be at most 32 and the actuall alignment is rounded by destWidth, so this is not a fixed number, do we still need set a fix number in soc level?

Comment (teburd) [reply tests/drivers/adc/adc_dma/boards/frdm_k64f.overlay:15]: Ok I see the problem now, yes so in this case dma-buf-alignment isn't really a static property of the dma controller, and that's fine. It shouldn't use that property at all then.

However... a magic alignment of 32 bytes isn't that great either in this case. Is 32 byte alignment always going to be used for NXP eDMA based devices with ADC? Perhaps adding an adc-buf-alignment property in this case is more appropriate.

Either way this information of needing a specific alignment for a buffer for this device should be encoded in some place besides in a test case. Removing my -1 as I think this needs to be decided by the ADC maintainer not by myself though.

Comment (hakehuang) [reply tests/drivers/adc/adc_dma/boards/frdm_k64f.overlay:15]: @teburd yes dma-buf-alignment = <4>, I updated.

**Comment by hakehuang** [comment]
the CI check failure is fixed by https://github.com/zephyrproject-rtos/zephyr/pull/50014

File: tests/drivers/adc/adc_dma/boards/frdm_k82f.overlay (hunk 4 above)
Comment (hakehuang) [reply tests/drivers/adc/adc_dma/boards/frdm_k82f.overlay]: @teburd  can you help to review and approve? Thanks

**Comment by mmahadevan108** [comment]
@stephanosio Can you help review this PR one more time to see if it addresses your concerns.

**Review by mmahadevan108 — APPROVED** [review]

**Comment by dleach02** [comment]
@teburd if you are good with the state can you approve?

**Comment by teburd** [comment]
@dleach02 @hakehuang I don't think the dma-buf-alignment property is the appropiate place in this PR (on the dma node) for this information in this case. Instead perhaps in the adc dts bindings an optional dma-buf-alignment would be appropriate here as its really the combination of adc + edma and not something that applies to all edma usages.

**Comment by hakehuang** [comment]
> @dleach02 @hakehuang I don't think the dma-buf-alignment property is in the appropriate place in this PR (on the dma node) for this information in this case. Instead perhaps in the adc dts bindings an optional dma-buf-alignment would be appropriate here as its really the combination of adc + edma and not something that applies to all edma usages.

@teburd  if we add this to adc then for every dma enabled peripheral we need add so, which means we have module inter-depency.  Could we accept this PR as a fast solution for this release, and let's discuss a better way to handle the DMA buffer in all peripheral controller side?

**Review by teburd — APPROVED** [review]

**Comment by teburd** [comment]
Sure, I'm approving today for this release, given the promise that we can further discuss this as I think its important. Not just for NXP and eDMA either!

**Comment by hakehuang** [comment]
> Sure, I'm approving today for this release, given the promise that we can further discuss this as I think its important. Not just for NXP and eDMA either!

Thanks @teburd 
I add one discussion for this purpose. we can discussion ideas with it. 

https://github.com/zephyrproject-rtos/zephyr/issues/50293

**Pull request merged**